### PR TITLE
added native MPL support and handling of outliers

### PR DESCRIPTION
**Pull request by fraseyboo**
These changes should improve a few things

### CODE REVIEW

File: GUI.py [hunk 1]
```diff
@@ -10,23 +10,66 @@
 from .fit import *
 import configparser
 
-language_path = os.path.realpath(__file__)[:-6] + "language\\"
```
Comment (fraseyboo) [inline GUI.py:13]: Using backslashes is only for Windows, I changed this to be platform independent

File: GUI.py [hunk 2]
```diff
@@ -10,23 +10,66 @@
...
+
+from PyQt5 import QtCore, QtGui, QtWidgets
+
+from matplotlib.backends.backend_qt5agg import FigureCanvasQTAgg, NavigationToolbar2QT as NavigationToolbar
+from matplotlib.figure import Figure
+
+
+
+
+try:
```
Comment (fraseyboo) [inline GUI.py:29]: This Latex-ifys the plot so it's more MATLAB-esque

File: GUI.py [hunk 3]
```diff
@@ -10,23 +10,66 @@
...
+        "text.usetex": True,
+        "font.family": "DejaVu Sans",
+        "font.serif": ["Palatino"],
+    })
+
+except:
+    print('Could not Latex-ify labels ')
+
+
+class MplCanvas(FigureCanvasQTAgg):
```
Comment (fraseyboo) [inline GUI.py:40]: Defining the new matplotlib canvas with support for residuals

File: GUI.py [hunk 4]
```diff
@@ -10,23 +10,66 @@
...
+
+language_path = os.path.realpath(__file__)[:-6] + "language" + os.sep
+
 class Mainwindow(QtWidgets.QMainWindow, Ui_MainWindow):
     def __init__(self, parent=None, rawVariables=None):
         super(Mainwindow, self).__init__(parent)
         self.setupUi(self)
         self.show()
 
-        # 设置图像框
```
Comment (fraseyboo) [inline GUI.py:22]: Removing the old QtGraph code

File: GUI.py [hunk 5]
```diff
@@ -10,23 +10,66 @@
...
-        self.graphWidget.showGrid(x=True, y=True)
+        self.show_residuals = True
+
+        # # 设置图像框
         self.Xlabel = None
         self.Ylabel = None
 
+        self.sc = MplCanvas(self, width=5, height=4, dpi=100, show_residuals=self.show_residuals)
+        self.sc.axes.grid(True)
+        self.toolbar = NavigationToolbar(self.sc, self)
```
Comment (fraseyboo) [inline GUI.py:67]: Adding an interactive toolbar for easier navigation and tweaking of the plots

File: GUI.py [hunk 6]
```diff
@@ -52,7 +95,7 @@ def __init__(self, parent=None, rawVariables=None):
         # 设置翻译
         self.translator = QtCore.QTranslator()
         config = configparser.ConfigParser()
-        config.read(os.path.expanduser('~') + '\curvefitting.ini')
+        config.read(os.path.expanduser('~') + os.sep + 'curvefitting.ini')
```
Comment (fraseyboo) [inline GUI.py:98]: for cross-platform support

File: GUI.py [hunk 7]
```diff
@@ -153,11 +196,16 @@ def plot(self):
...
-                pos = [{'pos': [x[i], y[i]]} for i in range(len(x))]
-                scatter.setData(pos)
+
+                self.full_x = x
+                self.full_y = y
+
+                self.sc.axes.clear()
+                self.sc.axes.scatter(x, y, s=5, c='k', label='Raw data', picker=True, pickradius=1)
+                self.outliers = np.zeros(x.shape).astype(bool)
+                self.sc.mpl_connect('pick_event', self.outlier_pick)
```
Comment (fraseyboo) [inline GUI.py:206]: Used to connect the plot to an update refresh that will select the point in close proximity and re-run the fit.

File: GUI.py [hunk 8]
```diff
@@ -199,9 +247,10 @@ def findfitmod(self):
             value = 0
         else:
             value = self.findcombo[str(index)].currentIndex()
+
         self.fitmod = (index, value)
 
-    def goodfit(self):
+    def goodfit(self, outliers=None):
```
Comment (fraseyboo) [inline GUI.py:253]: Added in support to exclude marked points from the fit

File: GUI.py [hunk 9]
```diff
@@ -224,12 +280,9 @@ def goodfit(self):
     def goodfitback(self):
         # 把fit操作放在一个线程里，结束后触发的恢复按钮状态的操作
         if self.successfit:
+
+
             try:
-                pen = pg.mkPen(color='b', width=4)
```
Comment (fraseyboo) [inline GUI.py:228]: Removing old QtGraph code

File: GUI.py [hunk 10]
```diff
@@ -238,11 +291,82 @@ def goodfitback(self):
             except Exception as e:
                 self.messege(repr(e))
         else:
+
             self.messege(self.e)
+            pass
+
+        self.update_figure()
```
Comment (fraseyboo) [inline GUI.py:298]: Calling the plot update as a function rather than within the callback

File: GUI.py [hunk 11]
```diff
@@ -308,9 +432,7 @@ def savefigure(self):
                                                                    os.getcwd(),
                                                                    "Portable Network Graphics(*.png);;Joint Photographic Group(*.jpg);;All Files(*)")
         if fileName:
-            ex = pyqtgraph.exporters.ImageExporter(self.graphWidget.plotItem)
-            ex.parameters()['width'] = 2000
-            ex.export(fileName)
+            self.sc.axes.figure.savefig(fileName, dpi=300)
```
Comment (fraseyboo) [inline GUI.py:435]: Using Matplotlib's native exporter

File: GUI.py [hunk 12]
```diff
@@ -386,8 +510,9 @@ def __init__(self, myui):
     def run(self):
         self.ui.successfit = False
         try:
-            self.ui.p, self.ui.para, self.ui.para_names = fit(self.ui.x, self.ui.y, mod=self.ui.fitmod)
+            self.ui.p, self.ui.para, self.ui.para_names = fit(self.ui.x.squeeze(), self.ui.y.squeeze(), mod=self.ui.fitmod)
```
Comment (fraseyboo) [inline GUI.py:513]: Squeeze appears necessary when autofit is disabled, I believe the return of a manual fit is not a 1D array as expected, this keeps it generic as a hot fix but should be investigated further

File: __init__.py [hunk 13]
```diff
@@ -11,10 +11,10 @@
 from .GUI import Mainwindow
 import configparser
 
-if not os.path.exists(os.path.expanduser('~')+'\curvefitting.ini'):
+if not os.path.exists(os.path.expanduser('~') + os.sep + 'curvefitting.ini'):
```
Comment (fraseyboo) [inline __init__.py:14]: Cross-platform change

File: __init__.py [hunk 14]
```diff
@@ -11,10 +11,10 @@
 from .GUI import Mainwindow
 import configparser
 
-if not os.path.exists(os.path.expanduser('~')+'\curvefitting.ini'):
+if not os.path.exists(os.path.expanduser('~') + os.sep + 'curvefitting.ini'):
     config = configparser.ConfigParser()
     config['DEFAULT'] = {"language": "en", "autofit": "False"}
-    with open(os.path.expanduser('~')+'\curvefitting.ini', 'w') as configfile:
+    with open(os.path.expanduser('~')+ os.sep + 'curvefitting.ini', 'w') as configfile:
```
Comment (fraseyboo) [inline __init__.py:17]: Ditto.

File: fit.py [hunk 15]
```diff
@@ -62,7 +62,8 @@ def give_reflect(x, y, f, para, para_names, mod, language):
              {"zh_CN": "{}项幂：", "en": " Power{}:"},
              {"zh_CN": "{}{}阶有理数：", "en": " Rat{}{}:"},
              {"zh_CN": "{}项正弦：", "en": " Sin{}:"},
-             {"zh_CN": "威布尔：", "en": "Weibull:"}]
+             {"zh_CN": "威布尔：", "en": "Weibull:"},
+             {"zh_CN": "威布尔：", "en": "Lorentzian:"}]
```
Comment (fraseyboo) [inline fit.py:66]: Attempting to add a new fit function, I don't have access to the ui code though.

**Comment by fansion314** [comment]
Thank you for helping me improve this project! I'm sorry I didn't see this message earlier. I did this project last year when I was a high school student. So what I did was not very well. But I'm trying to improve it though.

**Comment by fraseyboo** [comment]
I actually think you did a great job with this code, it certainly shows a lot of promise and I learned a lot from it. The only real issue I found was to do with the paths you set with the config file which only really affects Mac and Linux users. I then just switched the Qt graph canvas to use a Matplotlib canvas and added support for excluding outliers.

As far as I can tell there isn't really a good implementation of the curve fitting toolbox done outside MATLAB and now with my changes this program feels like a viable alternative. If you want I can continue to form pull requests to add in other features, it'd be cool to get 2D function fitting working and add in a few functions that are commonly used in my field.

I think the one thing that's stopping me from developing this further is that the Qt GUI code is a little hard to interpret, I assume you used Qt Designer for the GUI so maybe with the source code it'd be easier to add in the necessary list items & buttons to expand the functionality.

**Comment by fansion314** [comment]
Sorry, I couldn't find the original ui file on my current computer. I will look for it on my old computer. If I find it, I will add it latter.
Moreover, the fitting algorithm is to call scipy.optimize directly. I think it will be better if we can implement a more efficient algorithm.
For the next work, do you think it would be better to implement it with C++ Qt?

**Comment by fraseyboo** [comment]
I think the flexibility of scipy.optimize suits the program fine for the time being, the only issue I’ve  encountered is with it reaching a maximum number of iterations. The Matplotlib canvas is already a little heavy to plot so I don’t know if the speed boost would be noticeable. 

It’d be useful to get the GUI files as it’s a little difficult to modify some parts of the program without them.

Moving to C++ would be a serious undertaking for me (I’m not great at it), it’s something I’d be willing to learn but I don’t know how helpful my contributions would be.

**Comment by fansion314** [comment]
I added all the UI files. But the default language is Chinese, which may cause you trouble. If you want, you can change it into English. Please ignore the childish 'copyright' on the about page.

**Pull request merged**